upstairs/src/lib.rs

```diff
@@ -2607,6 +2411,186 @@ impl BlockIO for Guest {
+}
+
+pub struct GuestIoHandle {
+    /// Queue to receive new blockreqs
+    req_rx: mpsc::Receiver<BlockReq>,
+
+    /// Guest IO and bandwidth limits
+    limits: Arc<std::sync::Mutex<GuestLimits>>,
+
+    /// `BlockReq` that is at the head of the queue
+    ///
+    /// If a `BlockReq` was pulled from the queue but couldn't be used due to
+    /// IOP or bandwidth limiting, it's stored here instead (and we check this
+    /// before awaiting the queue).
+    req_head: Option<BlockReq>,
+
+    /// Are we currently IOP or bandwidth limited?
+    ///
+    /// If so, we don't return anything in `recv()`
+    req_limited: bool,
+
+    /// Current number of IOP tokens
+    iop_tokens: usize,
+
+    /// Current backpressure (shared with the `Guest`)
+    backpressure_us: Arc<AtomicU64>,
+
+    /// Backpressure configuration, as a starting point and max delay
+    backpressure_config: BackpressureConfig,
+
+    /// Bandwidth tokens (in bytes)
+    bw_tokens: usize,
+
+    /// Active work from the guest
+    ///
+    /// When the crucible listening task has noticed a new IO request, it
+    /// will pull it from the reqs queue and create an GuestWork struct
+    /// as well as convert the new IO request into the matching
+    /// downstairs request(s). Each new GuestWork request will get a
+    /// unique gw_id, which is also the index for that operation into the
+    /// hashmap.
+    ///
+    /// It is during this process that data will encrypted. For a read, the
+    /// data is decrypted back to the guest provided buffer after all the
+    /// required downstairs operations are completed.
+    guest_work: GuestWork,
+
+    /// Log handle, mainly to pass it into the [`Upstairs`]
+    log: Logger,
+}
+
+impl GuestIoHandle {
+    /// Leak IOPs tokens
+    pub fn leak_iop_tokens(&mut self, tokens: usize) {
+        self.iop_tokens = self.iop_tokens.saturating_sub(tokens);
+        self.req_limited = false;
+    }
+
+    /// Leak bytes from bandwidth tokens
+    pub fn leak_bw_tokens(&mut self, bytes: usize) {
+        self.bw_tokens = self.bw_tokens.saturating_sub(bytes);
+        self.req_limited = false;
+    }
+
+    /// Listen for new work
+    ///
+    /// This will wait forever if we are currently IOP / BW limited; otherwise,
+    /// it will return the next value from the `BlockReq` queue.
+    async fn recv(&mut self) -> BlockReq {
+        let req = if self.req_limited {
+            futures::future::pending().await
+        } else if let Some(req) = self.req_head.take() {
+            req
+        } else if let Some(req) = self.req_rx.recv().await {
+            // NOTE: once we take this req from the queue, we must be cancel
+            // safe!  In other words, we cannot yield until either (1) returning
+            // the req or (2) storing it in self.req_head for safe-keeping.
+            req
+        } else {
+            futures::future::pending().await
+        };
+
+        // Check if we can consume right away
```

Comment:
The meat of this method appears to be pretty much the same as it was before.
Thanks github for not making that easy to determine.

Reply:
Yes, that's correct – once we get the request off the queue, the logic should be the same.